JetBrains: Make the plugin compatible with 3.42.0 and below

vdavid opened the pull request:
Fixes https://github.com/sourcegraph/sourcegraph/issues/41897

The problem is that Sourcegraph version 3.42.0 and the ones below didn't have `patternType: standard`. ([source](https://docs.sourcegraph.com/CHANGELOG#3-43-0))

The fix gets the version from Sourcegraph at the initial auth (I managed to achieve this in an elegant way without an additional request), and if the version is affected, it falls back to `patternType: literal`. I didn't find the difference between these two, but they should be the same for most purposes, if not all.

I made sure that the standalone version remains a useful testing tool, so I've updated it in all the ways that made sense.

For the reviewer: going commit by commit might help, I tried my best to leave nice commit messages.

Note: It handles custom builds (like "174153_2022-09-26_f120e7801981") well as long as they are newer than 3.42.0 (it just considers all non-standard versions as newer).

## Test plan

The test situations are tricky to create ([this helps](https://docs.sourcegraph.com/admin/deploy/docker-single-container)), so I've tested in this in this [5-min Loom](https://www.loom.com/share/2f7620d411d24109af8c08a471e862e4).

Also added the note to the overlay that I mentioned in the video:

![CleanShot 2022-09-26 at 16 57 10@2x](https://user-images.githubusercontent.com/2552265/192310687-31341d4c-1f1a-4be2-a2b3-d810f3c69488.png)

## App preview:

- [Web](https://sg-web-dv-jetbrains-make-compatible-with.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-ddzcqikuah.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

vdavid commented:
I've noticed a bug in this fix. If I start the plugin with my wifi off, wait for the error message to come up, then re-enable the wifi, then the error message is stuck unless I change the plugin settings or reopen the project. This is not nice but it's not a trivial fix because we don't have a straightforward way yet to externally trigger a search in the webview.

philipp-spiess commented on client/jetbrains/webview/src/bridge-mock/call-java-mock.ts:
```diff
@@ -156,6 +167,11 @@ export function setDarkMode(value: boolean): void {
 }
 
+function setOverlay(content: string | null): void {
+    webviewOverlay.style.visibility = content !== null ? 'visible' : 'hidden'
```
It's probably irrelevant because of the `position: absolute` but it's usually faster to use `display: none` over `visibility: hidden` since visibility will retain the box's dimensions so the content will still need to be rendered in the layout steps. With `position: absolute` and a hard coded size this is probably a no-op though (since the position won't affect the rest of the layout).

philipp-spiess commented on client/jetbrains/webview/src/sourcegraph-api-access/api-gateway.ts:
```diff
@@ -1,17 +1,63 @@
+                canSignOut
+            }
+            viewerCanAdminister
+            tags
```
Do we need all of these fields for the user? It looks like we may be over-fetching quite a bit here but not sure! If this is needed since it is used somewhere else (it seems to be shared via the `AuthenticatedUser`?), we should consider sharing the GraphQL query as a fragment so that we only need to define it once in our schema.

vdavid replied:
> Do we need all of these fields for the user?

In the plugin, we don't, but we pass the data to the search box, and for that, it needs to be a fully qualified `AuthenticatedUser`. 😕 

> fragments

Before my changes, we used [this query](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@dv/jetbrains-make-compatible-with-3.41.0/-/blob/client/shared/src/auth.ts?L5%3A31=), but I had to add `site { productVersion } to it, and TBH I didn't know about fragments 😅. I don't want to mess with it now, though, so I've just added a TODO that deduplication would be nice using fragments.

philipp-spiess commented on client/jetbrains/webview/src/search/App.tsx:
```diff
@@ -56,6 +57,16 @@ function fetchStreamSuggestionsWithStaticUrl(query: string): Observable<SearchMa
+    const [major, minor] = backendVersion.split('.').map(Number)
+    // SearchPatternType.standard is not supported by versions before 3.43.0
+    return (major < 3 || (major === 3 && minor < 43)) ? SearchPatternType.literal : SearchPatternType.standard
+}
```
Nice!

philipp-spiess reviewed (approved):
Great work! These changes LGTM.

> 
I've noticed a bug in this fix. If I start the plugin with my wifi off, wait for the error message to come up, then re-enable the wifi, then the error message is stuck unless I change the plugin settings or reopen the project. This is not nice but it's not a trivial fix because we don't have a straightforward way yet to externally trigger a search in the webview.

Hmm I see. This seems unrelated to the initial issue that you wanted to fix though so perhaps we should prioritize it different (i.e put on the backlog for now)?

vdavid commented:
@philipp-spiess 

> This seems unrelated to the initial issue that you wanted to fix though so perhaps we should prioritize it different (i.e put on the backlog for now)?

Unfortunately, it was caused by this change: the UX in these cases got worse after my changes.
Today I fixed it, though. Once I was there, I also fixed a previously existing problem related to a lack of internet connection. I demo it all [here in two minutes](https://www.loom.com/share/ca8884d44144418ea729d4871d76df13).
Please re-review the last three commits. Thank you! :)

philipp-spiess commented:
@vdavid Sorry I overlooked your ping and since it was already reviewed this did not show up in my list of PRs to review :(
